compiler/qsc_eval/src/lib.rs

```diff
@@ -1448,6 +1462,31 @@ impl State {
             span,
         }
     }
+
+    fn counting_call(&mut self, name: &str, arg: Value) -> Value {
+        let callable = if let Value::Closure(closure) = arg {
+            (closure.id, closure.functor)
+        } else {
+            arg.unwrap_global()
+        };
+        match name {
+            "StartCountingOperation" | "StartCountingFunction" => {
+                self.call_counts.insert(callable, 0);
+                Value::unit()
+            }
+            "StopCountingOperation" | "StopCountingFunction" => {
+                let count = self.call_counts.remove(&callable).unwrap_or(-1);
+                Value::Int(count)
+            }
+            _ => panic!("unknown counting call"),
+        }
+    }
+
+    fn increment_call_count(&mut self, callee_id: StoreItemId, functor: FunctorApp) {
```

Comment:
Did you test if this has any notable overhead - esp. something like resource estimation where call counts could be very high with no actual quantum simulation overhead? I can think of some trivial optimizations if needed, but if no noticable impact then doesn't matter.

Reply:
Good question. I ran the Deutsch-Jozsa benchmark on main, then compared it to this branch in two ways:

1. No operations counted, so just checking the overhead of the checks when no callables are in the tracking hashmap.
2. With `X`, `H`, `CX`, `CNOT`, `Controlled X` all registered, such that 5 operations are in the tracking hashmap that are checked for every call and sometimes hit.

For both versions, the benchmark showed no change in performance (behavior was within noise, such that 1500 iterations ran and some were faster and some were slower, but all by less than +/- 1%). That matches intuition: there would have to be a very high number of callables registered in the hashmap for checking membership to trigger any measurable performance difference.